Source/WebCore/editing/markup.cpp

```diff
@@ -1345,7 +1346,16 @@ static ALWAYS_INLINE ExceptionOr<Ref<DocumentFragment>> createFragmentForMarkup(
     auto fragment = mode == DocumentFragmentMode::New ? DocumentFragment::create(document.get()) : document->documentFragmentForInnerOuterHTML();
     ASSERT(!fragment->hasChildNodes());
     if (document->isHTMLDocument()) {
-        fragment->parseHTML(markup, &contextElement, parserContentPolicy);
+        bool parsedUsingFastPath = tryFastParsingHTMLFragment(markup, document, fragment, contextElement, parserContentPolicy);
+        if (parsedUsingFastPath) {
+#if ASSERT_ENABLED
+            // As a sanity check for the fast-path, create another fragment using the full parser and compare the results.
+            auto referenceFragment = DocumentFragment::create(document);
+            referenceFragment->parseHTML(markup, &contextElement, parserContentPolicy);
+            ASSERT(serializeFragment(fragment, SerializedNodes::SubtreesOfChildren) == serializeFragment(referenceFragment, SerializedNodes::SubtreesOfChildren));
+#endif
+        } else
+            fragment->parseHTML(markup, &contextElement, parserContentPolicy);
```

Comment:
Could this all go into the `fragment->parseHTML` function, not here? The abstraction is "faster way to parse", not "two different ways to parse", so I think one level deeper.

Reply:
The problem is that the fast path can start parsing and populating the fragment, then encounter something it can't parse and give up. When it gives up we currently destroy the fragment and reconstruct a new one before calling `fragment->parseHTML()`. This reconstruction makes what you suggest not possible so we'd need another approach. I already tried calling `removeChildren()` on the fragment but that turned out to be too expensive. I need to find a cheaper way.